server/src/main/java/io/deephaven/server/notebook/FilesystemStorageServiceGrpcImpl.java

```diff
@@ -0,0 +1,210 @@
+import io.deephaven.proto.backplane.grpc.FetchFileResponse;
+import io.deephaven.proto.backplane.grpc.FileInfo;
+import io.deephaven.proto.backplane.grpc.FileKind;
+import io.deephaven.proto.backplane.grpc.ListItemsRequest;
+import io.deephaven.proto.backplane.grpc.ListItemsResponse;
+import io.deephaven.proto.backplane.grpc.MoveItemRequest;
+import io.deephaven.proto.backplane.grpc.MoveItemResponse;
+import io.deephaven.proto.backplane.grpc.SaveFileRequest;
+import io.deephaven.proto.backplane.grpc.SaveFileResponse;
+import io.deephaven.proto.backplane.grpc.StorageServiceGrpc;
+import io.deephaven.server.session.SessionService;
+import io.grpc.stub.StreamObserver;
+
+import javax.inject.Inject;
+import javax.inject.Singleton;
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.nio.file.FileAlreadyExistsException;
+import java.nio.file.FileSystems;
+import java.nio.file.Files;
+import java.nio.file.NoSuchFileException;
+import java.nio.file.Path;
+import java.nio.file.PathMatcher;
+import java.nio.file.Paths;
+import java.nio.file.StandardOpenOption;
+import java.util.Optional;
+import java.util.regex.PatternSyntaxException;
+import java.util.stream.Stream;
+
+@Singleton
+public class FilesystemStorageServiceGrpcImpl extends StorageServiceGrpc.StorageServiceImplBase {
+    private static final Logger log = LoggerFactory.getLogger(FilesystemStorageServiceGrpcImpl.class);
+
+    private static final String STORAGE_PATH =
+            Configuration.getInstance().getStringWithDefault("storage.path", "<workspace>/storage")
+                    .replace("<workspace>", Configuration.getInstance().getWorkspacePath());
+
+    private final Path root = Paths.get(STORAGE_PATH).normalize();
+    private final SessionService sessionService;
+
+    @Inject
+    public FilesystemStorageServiceGrpcImpl(SessionService sessionService) {
+        this.sessionService = sessionService;
+        try {
+            Files.createDirectories(root);
+        } catch (IOException e) {
+            throw new UncheckedIOException("Failed to initialize storage", e);
+        }
+    }
+
+    private Path resolveOrThrow(String relativePath) {
+        Path resolved = root.resolve(relativePath).normalize();
+        if (resolved.startsWith(root)) {
+            return resolved;
+        }
+        throw GrpcUtil.statusRuntimeException(Code.INVALID_ARGUMENT, "Invalid path: " + relativePath);
+    }
+
+    @Override
+    public void listItems(ListItemsRequest request, StreamObserver<ListItemsResponse> responseObserver) {
+        GrpcUtil.rpcWrapper(log, responseObserver, () -> {
+            sessionService.getCurrentSession();
+
+            ListItemsResponse.Builder builder = ListItemsResponse.newBuilder();
+            PathMatcher matcher = request.hasFilterGlob() ? makeMatcher(request.getFilterGlob()) : ignore -> true;
+            Path dir = resolveOrThrow(request.getPath());
+            try (Stream<Path> list = Files.list(dir)) {
```

Comment:
Also, we may not need to make sure the glob is safe w/ newDirectoryStream - I think it only resolves in that directory (but not 100% sure).

Reply:
Yes, that saves a line here, but it still has to be closed.

We want the glob to only resolve in that directory, so that would be good. I wouldnt mind sending an error back still if the user attempted to do a glob that doesnt actually make sense.

Reply:
On second thought, to get good parse errors, i'm going to split the difference a bit, and keep parsing the pattern in our own code, but pass it to newDirectoryStream. It does still save us a matcher check and a method reference.

Reply:
Nope, this doesn't work - newDirectoryStream will not pass the actual item being streamed, but first resolved it against the parent. Technically we could probably prefix all globs with the parent path (must be the same absolute vs relative as the stream itself will use) or prefix all globs with `**`, but it seems cleaner to just match patterns as they were written, at the cost of having to make an iterable.